aiohttp/http_websocket.py

```diff
@@ -304,134 +307,150 @@ def feed_data(self, data: bytes) -> Tuple[bool, bytes]:
-
-                    if opcode == WSMsgType.CONTINUATION:
-                        assert self._opcode is not None
-                        opcode = self._opcode
-                        self._opcode = None
-
-                    self._partial.extend(payload)
-                    if self._max_msg_size and len(self._partial) >= self._max_msg_size:
-                        raise WebSocketError(
-                            WSCloseCode.MESSAGE_TOO_BIG,
-                            "Message size {} exceeds limit {}".format(
-                                len(self._partial), self._max_msg_size
-                            ),
-                        )
-
-                    # Decompress process must to be done after all packets
-                    # received.
-                    if compressed:
-                        assert self._decompressobj is not None
-                        self._partial.extend(_WS_DEFLATE_TRAILING)
-                        payload_merged = self._decompressobj.decompress_sync(
-                            self._partial, self._max_msg_size
-                        )
-                        if self._decompressobj.unconsumed_tail:
-                            left = len(self._decompressobj.unconsumed_tail)
-                            raise WebSocketError(
-                                WSCloseCode.MESSAGE_TOO_BIG,
-                                "Decompressed message size {} exceeds limit {}".format(
-                                    self._max_msg_size + left, self._max_msg_size
-                                ),
-                            )
-                    else:
-                        payload_merged = bytes(self._partial)
-
-                    self._partial.clear()
-
-                    if opcode == WSMsgType.TEXT:
-                        try:
-                            text = payload_merged.decode("utf-8")
-                            self.queue.feed_data(WSMessage(WSMsgType.TEXT, text, ""))
-                        except UnicodeDecodeError as exc:
-                            raise WebSocketError(
-                                WSCloseCode.INVALID_TEXT, "Invalid UTF-8 text message"
-                            ) from exc
-                    else:
-                        self.queue.feed_data(
-                            WSMessage(WSMsgType.BINARY, payload_merged, ""),
-                        )
+            return
+
+        # previous frame was non finished
+        # we should get continuation opcode
+        if self._partial and not is_continuation:
+            raise WebSocketError(
+                WSCloseCode.PROTOCOL_ERROR,
+                "The opcode in non-fin frame is expected "
+                "to be zero, got {!r}".format(opcode),
+            )
 
-        return False, b""
+        if is_continuation:
+            assert self._opcode is not None
+            opcode = self._opcode
+            self._opcode = None
+
+        self._partial.extend(payload)
```

Comment:
This is where most of the overhead is. When there are many small frames we end up making lots of copies that we do not need to make

Reply:
https://stackoverflow.com/questions/40004517/why-does-bytearray-obj-extendbytes-differ-from-bytearray-obj-bytes